service/header/sync.go

```diff
@@ -88,40 +146,66 @@ func (s *Syncer) finishSync() {
 	atomic.StoreUint64(&s.inProgress, 0)
 }
 
-// Validate implements validation of incoming Headers and stores them if they are good.
-func (s *Syncer) Validate(ctx context.Context, p peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
-	header, err := UnmarshalExtendedHeader(msg.Data)
+func (s *Syncer) validateMsg(ctx context.Context, p peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+	maybeHead, err := UnmarshalExtendedHeader(msg.Data)
 	if err != nil {
-		log.Errorw("unmarshalling ExtendedHeader received from the PubSub",
+		log.Errorw("unmarshalling header received from the PubSub",
 			"err", err, "peer", p.ShortString())
 		return pubsub.ValidationReject
 	}
 
-	// if syncing is still in progress - just ignore the new header as
-	// Syncer will fetch it after anyway, but if syncer is done, append
-	// the header.
-	if !s.IsSyncing() {
-		err := s.store.Append(ctx, header)
-		if err != nil {
-			log.Errorw("appending store with header from PubSub",
-				"hash", header.Hash().String(), "height", header.Height, "peer", p.ShortString())
-			// TODO(@Wondertan): We need to be sure that the error is actually validation error.
-			//  Rejecting msgs due to storage error is not good, but for now that's fine.
-			return pubsub.ValidationReject
-		}
+	return s.validate(ctx, p, maybeHead)
+}
 
-		// we are good to go
+// validate implements validation of incoming Headers and stores them if they are good.
+func (s *Syncer) validate(ctx context.Context, p peer.ID, maybeHead *ExtendedHeader) pubsub.ValidationResult {
+	localHead, err := s.store.Head(ctx)
+	if err != nil {
+		log.Errorw("getting local head", "err", err)
+		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
+	}
+
+	if maybeHead.Height <= localHead.Height {
+		log.Warnw("rcvd known header", "from", p.ShortString())
+		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
+	}
+
+	// validate that +2/3 of subjective validator set signed the commit and if not - reject
+	err = Verify(localHead, maybeHead)
+	if err != nil {
+		// our subjective view can be far from objective network head(during sync), so we cannot be sure if header is
+		// 100% invalid due to outdated validator set, thus ValidationIgnore and thus 'possibly malicious'
+		log.Warnw("rcvd possibly malicious header",
+			"err", err, "hash", maybeHead.Hash(), "height", maybeHead.Height, "from", p.ShortString())
+		return pubsub.ValidationIgnore
+	}
+
+	if maybeHead.Height > localHead.Height+1 {
+		// we might be missing some headers, so save verified header to pending cache
+		s.pending.Add(maybeHead)
```

Comment:
Could this be gamed by sending a node a bunch of headers that just keep growing and growing `pending`?

Reply:
Like this question! No, you can't because Verify above ensures that each header coming to pending should have 2/3 power overlap with the local subjective head. One thing I am also considering rn, is to check against the objective head we get from a trusted peer.

Reply:
Can you document your response?